app/api/items/save_grade.go

```diff
@@ -0,0 +1,275 @@
+	"github.com/France-ioi/AlgoreaBackend/app/token"
+)
+
+func (srv *Service) saveGrade(w http.ResponseWriter, r *http.Request) service.APIError {
+	requestData := SaveGradeRequest{store: srv.Store, publicKey: srv.TokenConfig.PublicKey}
+
+	var err error
+	if err = render.Bind(r, &requestData); err != nil {
+		return service.ErrInvalidRequest(err)
+	}
+
+	user := srv.GetUser(r)
+	if err = user.Load(); err == database.ErrUserNotFound {
+		return service.InsufficientAccessRightsError
+	}
+	service.MustNotBeError(err)
+
+	apiError := service.NoError
+	if apiError = checkSaveGradeTokenParams(user, &requestData); apiError != service.NoError {
+		return apiError
+	}
+
+	var validated, keyObtained bool
+	err = srv.Store.InTransaction(func(store *database.DataStore) error {
+		var hasAccess bool
+		var reason error
+		hasAccess, reason, err = store.Items().CheckSubmissionRights(requestData.TaskToken.Converted.LocalItemID, user)
+		service.MustNotBeError(err)
+
+		if !hasAccess {
+			apiError = service.ErrForbidden(reason)
+			return nil // commit! (CheckSubmissionRights() changes the DB sometimes)
+		}
+
+		validated, keyObtained = saveGradingResultsIntoDB(store, user, &requestData)
+		return nil
+	})
+	if apiError != service.NoError {
+		return apiError
+	}
+	service.MustNotBeError(err)
+
+	if validated && requestData.TaskToken.Converted.AccessSolutions != nil && !(*requestData.TaskToken.Converted.AccessSolutions) {
+		requestData.TaskToken.AccessSolutions = formdata.AnythingFromString(`"1"`)
+	}
+	requestData.TaskToken.PlatformName = srv.TokenConfig.PlatformName
+	newTaskToken, err := requestData.TaskToken.Sign(srv.TokenConfig.PrivateKey)
+	service.MustNotBeError(err)
+
+	service.MustNotBeError(render.Render(w, r, service.CreationSuccess(map[string]interface{}{
+		"task_token":   newTaskToken,
+		"bValidated":   validated,
+		"bKeyObtained": keyObtained,
+	})))
+	return service.NoError
+}
+
+func saveGradingResultsIntoDB(store *database.DataStore, user *database.User,
+	requestData *SaveGradeRequest) (validated, keyObtained bool) {
+	const todo = "todo"
+	score := requestData.ScoreToken.Converted.Score
+	userAnswerID := requestData.ScoreToken.Converted.UserAnswerID
+
+	// TODO: handle validation in a proper way (what did he mean??)
+	if score > 99 {
+		validated = true
+	}
+	service.MustNotBeError(store.UserAnswers().ByID(userAnswerID).
+		Where("idUser = ?", user.UserID).
+		Where("idItem = ?", requestData.TaskToken.Converted.LocalItemID).
+		UpdateColumn(map[string]interface{}{
+			"sGradingDate": gorm.Expr("NOW()"),
+			"bValidated":   validated,
+			"iScore":       score,
+		}).Error())
+	// Build query to update users_items
+	// The iScore is set towards the end, so that the IF condition on
+	// sBestAnswerDate is computed before iScore is updated
+	columnsToUpdate := []string{
+		"nbTasksTried",
+		"sLastActivityDate",
+		"sBestAnswerDate",
+		"sLastAnswerDate",
+		"iScore",
+	}
+	values := []interface{}{
```

Comment:
Could you check that Michel's answer on time matches your implementation (I have the impression that some parts are missing, but maybe I've found the corresponding code yet): 

When the backend receives a score token : check the corresponding users_answers row, if users_answers.iScore is NULL, update users_answers.iScore (with the score in the score token), and set to NOW() users_answers.sGradingDate, {groups_attempts,users_items}.sBestAnswerDate (if the score was strictly better), {groups_attempts,users_items}.sLastAnswerDate
if users_answers.iScore is not NULL, then we had already received a score for this, so don't update anything

Reply:
We don't check if users_answers.iScore is null or not (and AlgoreaPlatform doesn't check that). Other things are all right.

Reply:
Should we check users_answers.iScore?

Reply:
Now it checks users_answers.iScore